File: ios/Packages/PadelCore/Sources/PadelCore/Model/Vector3.swift
import Foundation

/// Vector 3D inmutable en el marco de referencia del dispositivo.
public struct Vector3: Equatable, Sendable {
    public let x: Float
    public let y: Float
    public let z: Float

    public init(_ x: Float, _ y: Float, _ z: Float) {
        self.x = x
        self.y = y
        self.z = z
    }

    public static let zero = Vector3(0, 0, 0)

    public func magnitude() -> Float {
        (x * x + y * y + z * z).squareRoot()
    }

    public func dot(_ other: Vector3) -> Float {
        x * other.x + y * other.y + z * other.z
    }

    /// Vector unitario. Devuelve `zero` en vez de NaN si la magnitud es 0.
    public func normalized() -> Vector3 {
        let m = magnitude()
        guard m >= 1e-6 else { return .zero }
        return Vector3(x / m, y / m, z / m)
    }

    public static func + (lhs: Vector3, rhs: Vector3) -> Vector3 {
        Vector3(lhs.x + rhs.x, lhs.y + rhs.y, lhs.z + rhs.z)
    }

    public static func - (lhs: Vector3, rhs: Vector3) -> Vector3 {
        Vector3(lhs.x - rhs.x, lhs.y - rhs.y, lhs.z - rhs.z)
    }

    public static func * (lhs: Vector3, rhs: Float) -> Vector3 {
        Vector3(lhs.x * rhs, lhs.y * rhs, lhs.z * rhs)
    }

    public static prefix func - (value: Vector3) -> Vector3 {
        Vector3(-value.x, -value.y, -value.z)
    }
}

File: android/core/src/main/kotlin/com/risingpadel/core/model/Vector3.kt
package com.risingpadel.core.model

import kotlin.math.acos
import kotlin.math.sqrt

/** Vector 3D inmutable en el marco de referencia del dispositivo. */
data class Vector3(val x: Float, val y: Float, val z: Float) {

    fun magnitude(): Float = sqrt(x * x + y * y + z * z)

    fun dot(other: Vector3): Float = x * other.x + y * other.y + z * other.z

    /** Devuelve el vector unitario. Si la magnitud es 0 devuelve ZERO en vez de NaN. */
    fun normalized(): Vector3 {
        val m = magnitude()
        return if (m < 1e-6f) ZERO else Vector3(x / m, y / m, z / m)
    }

    operator fun plus(other: Vector3) = Vector3(x + other.x, y + other.y, z + other.z)

    operator fun minus(other: Vector3) = Vector3(x - other.x, y - other.y, z - other.z)

    operator fun times(scalar: Float) = Vector3(x * scalar, y * scalar, z * scalar)

    operator fun unaryMinus() = Vector3(-x, -y, -z)

    /** Ángulo en radianes con otro vector, en [0, PI]. */
    fun angleTo(other: Vector3): Float {
        val cos = normalized().dot(other.normalized()).coerceIn(-1f, 1f)
        return acos(cos)
    }

    companion object {
        val ZERO = Vector3(0f, 0f, 0f)
    }
}
